### PR TITLE
IA -> Add Response ByteCount & Wait(sec) implementation

### PR DESCRIPTION
**Pull request by DStavilaNI**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/niveristand-instrument-addon-custom-device/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Implementation of the Add Response - Wait(sec).vi
- Implementation of the Add Response - ByteCount.vi
- Implementation of the Test VIs

### Why should this Pull Request be merged?

- Implements the "Add Response API" and and corresponding tests

### What testing has been done?

- Implemented test VIs for the created API
- Testing on the local DEV machine has been done

### CODE REVIEW

**Comment by bogdanp-ni** [comment]
There are some tests where the "TestCase out" is unwired. Please wire them accordingly.

**Comment by bogdanp-ni** [comment]
For improving readability, please wire the same constant to these checks (as far as I understood, they should have the same value), instead od using two different ones (see below):

image

**Comment by phutanu** [comment]
Please remove the extra space at the end in the VI names for:
- Resolve Seconds Wait .vi
- Resolve ByteCount Wait .vi

**Comment by bogdanp-ni** [comment]
[Recommend] Wherever possible, I suggest scaling the constant arrays being used, so that all elements are visible.

**Comment by bogdanp-ni** [comment]
Parse Format String.vi -> this functionality is already present in the IA (Find Equation Variables.vi, in SE library). Please remove one of them, so that we don't have duplicated code.

**Comment by bogdanp-ni** [comment]
Parse Format String.vi -> typo here:
image

**Comment by phutanu** [comment]
Please adjust the names for the below marked inputs. Please use one of the below options:
1. "Item Ref In" and "Item Ref Out"
2. "Message In" and "Message Out"

image

**Comment by bogdanp-ni** [comment]
Resolve Response Linkaging.vi -> I must admit I'm having a hard time understanding the VI, mostly because of the multiple nested Case Structures. To improve readability, I would like you to make the following changes:
- the check for "empty Response" can be put in a separate subVI (throws error when empty), called right before Parse Format String, thus relying on the error chain to ensure correct execution
- a similar solution for checking that the array size from Parse Format String is 1
- the innermost functionality (see below) can be extracted in a subVI:
image

**Comment by bogdanp-ni** [comment]
Taking into consideration the fact that Add Response is linking variables with channel refnums (which is different from the UI logic), is there a way to utilize this scripting node (or some of its subVIs) directly in the UI part(SE lib)? In the long run, it will be very beneficial to remove as much duplication as possible, yielding a more lean and robust code.

**Review by bogdanp-ni — CHANGES_REQUESTED** [review]
Please see my comments above.

**Comment by phutanu** [comment]
> For improving readability, please wire the same constant to these checks (as far as I understood, they should have the same value), instead od using two different ones (see below):
> 
> image

Also please correct the typo in the constant names: "Re**s**ponse Parameters".

**Review by phutanu — CHANGES_REQUESTED** [review]
Please check my above comments.

**Comment by DStavilaNI** [comment]
> 
> 
> Please remove the extra space at the end in the VI names for:
> 
>     * Resolve Seconds Wait .vi
> 
>     * Resolve ByteCount Wait .vi

This is not recommended ?

**Comment by bogdanp-ni** [comment]
> > Please remove the extra space at the end in the VI names for:
> > ```
> > * Resolve Seconds Wait .vi
> > 
> > * Resolve ByteCount Wait .vi
> > ```
> 
> This is not recommended ?

I don't know about phutanu, so I only speak for myself now, but unless there's a specific reason for that extra space (I thought it was a typo), I definitely like to see it fixed.

**Comment by DStavilaNI** [comment]
> 
> 
> Resolve Response Linkaging.vi -> I must admit I'm having a hard time understanding the VI, mostly because of the multiple nested Case Structures. To improve readability, I would like you to make the following changes:
> 
>     * the check for "empty Response" can be put in a separate subVI (throws error when empty), called right before Parse Format String, thus relying on the error chain to ensure correct execution
> 
>     * a similar solution for checking that the array size from Parse Format String is 1
> 
>     * the innermost functionality (see below) can be extracted in a subVI:
>       image

Sure thing - I will refactor it to make it easier to understand

**Comment by DStavilaNI** [comment]
> 
> 
> > > Please remove the extra space at the end in the VI names for:
> > > ```
> > > * Resolve Seconds Wait .vi
> > > 
> > > * Resolve ByteCount Wait .vi
> > > ```
> > 
> > 
> > This is not recommended ?
> 
> I don't know about phutanu, so I only speak for myself now, but unless there's a specific reason for that extra space (I thought it was a typo), I definitely like to see it fixed.

Sure

**Comment by DStavilaNI** [comment]
> 
> 
> Taking into consideration the fact that Add Response is linking variables with channel refnums (which is different from the UI logic), is there a way to utilize this scripting node (or some of its subVIs) directly in the UI part(SE lib)? In the long run, it will be very beneficial to remove as much duplication as possible, yielding a more lean and robust code.

There are ways, but this would not be a simple copy&paste operation. I think extensive refactoring would be required in the SE and this would be a separate task/US.

**Comment by phutanu** [comment]
[Recommendation] VI naming.
In order to improve API code reuse and readability inside project libraries I would suggest not using shorting in VI names.
VI in question: "Check LC&RF Size Equality.vi"

**Review by phutanu — CHANGES_REQUESTED** [review]
Please revert "Remove Target from Node Path.vi" as I suspect it was submitted by accident. No change was made to it.

**Comment by phutanu** [comment]
> [Recommendation] VI naming.
> In order to improve API code reuse and readability inside project libraries I would suggest not using shorting in VI names.
> VI in question: "Check LC&RF Size Equality.vi"

For example this VI can be reused in the "Add Command" node. And that's why I would like it to have a verbose name.

File: Source/APIs/Scripting API/Instrument Addon Scripting API.lvlib [hunk 1]
```diff
@@ -85,10 +86,21 @@
 		<Item Name="Validate Instrument Addon Name.vi" Type="VI" URL="../SubVIs/Validate Instrument Addon Name.vi"/>
 		<Item Name="Validate Instrument Addon Name Is Not In Use.vi" Type="VI" URL="../SubVIs/Validate Instrument Addon Name Is Not In Use.vi"/>
 		<Item Name="Validate Instrument Addon Refnum.vi" Type="VI" URL="../SubVIs/Validate Instrument Addon Refnum.vi"/>
+		<Item Name="Resolve ByteCount Wait.vi" Type="VI" URL="../SubVIs/Resolve ByteCount Wait.vi"/>
+		<Item Name="Resolve Response Linkaging.vi" Type="VI" URL="../SubVIs/Resolve Response Linkaging.vi"/>
+		<Item Name="Resolve Seconds Wait.vi" Type="VI" URL="../SubVIs/Resolve Seconds Wait.vi"/>
+		<Item Name="Validate Message Refnum.vi" Type="VI" URL="../SubVIs/Validate Message Refnum.vi"/>
+		<Item Name="Check for Empty Response.vi" Type="VI" URL="../SubVIs/Check for Empty Response.vi"/>
+		<Item Name="Check for Duplicate Variables.vi" Type="VI" URL="../SubVIs/Check for Duplicate Variables.vi"/>
+		<Item Name="Check LC&amp;RF Size Equality.vi" Type="VI" URL="../SubVIs/Check LC&amp;RF Size Equality.vi"/>
```
Comment (phutanu) [inline Source/APIs/Scripting API/Instrument Addon Scripting API.lvlib:95]: Please update also the Scripting API.lvlib with the new VI name.

**Review by bogdanp-ni — APPROVED** [review]
I approve the changes. Regarding the refactor in SE lib, please add another task under this US, which you can address in another follow-up PR.

**Review by phutanu — APPROVED** [review]

**Pull request merged**